doc/tutorials/add_solver.rst

```diff
@@ -0,0 +1,331 @@
+
+- **iteration**
+
+This sampling strategy is for solvers that can be controlled using the maximum number of iterations performed.
+In this case, benchopt treats the solver as a black box and observes its behavior for different number of iterations.
+
+Therefore, the signature of the ``run`` method is ``run(self, n_iter)`` and its implementation resembles the snippet below.
+
+.. code-block:: python
+    :caption: solvers/mysolver.py
+
+    class Solver(BaseSolver):
+        ...
+        sampling_strategy = 'iteration'
+        ...
+
+        def run(self, n_iter):
+            # mysolver.solve is the black box solver; the names of its arguments
+            # may of course differ: you should adapt to your case!
+            beta = mysolver.solve(self.X, self.y, self.lmbd, n_iter=n_iter)
+
+            # store reference to the solution
+            self.beta = beta
+        ...
+
+- **tolerance**
+
+Similar to **iteration**, this sampling strategy is used for solver controlled by the tolerance on the optimization process.
+In this case, the signature of the ``run`` method is ``run(self, tol)`` and would be implemented as follows.
+
+.. code-block:: python
+    :caption: solvers/mysolver.py
+
+    class Solver(BaseSolver):
+        ...
+        sampling_strategy = 'tolerance'
+        ...
+
+        def run(self, n_iter):
+            # mysolver.solve is the black box solver; the names of its arguments
+            # may of course differ: you should adapt to your case!
+            beta = mysolver.solve(self.X, self.y, self.lmbd, tol=tol)
+
+            # store reference to the solution
+            self.beta = beta
+        ...
+
+- **callback**
+
+This strategy can be used when the solver is not a black box, but instead exposes the intermediate values the iterates.
+
+Here is a as snippet that illustrate how it could be implemented.
+
+.. code-block:: python
+    :caption: solvers/mysolver.py
+
+    class Solver(BaseSolver):
+        ...
+        sampling_strategy = 'callback'
+        ...
+
+        def run(self, callback):
+
+            while callback():
+                # do one iteration of the solver here:
+                beta = ...
+
+            # at the end of while loop, store reference to the solution
+            self.beta = beta
+        ...
+
+
+Getting the solver's results
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Finally, we define a ``get_result`` method that is used to pass the solver's result back to the objective.
+It must return a dictionary whose keys are the input arguments of ``Objective.evaluate_result``.
+
+In our case the input of ``Objective.evaluate_result`` is ``beta``, hence we return a dictionary with a single key ``"beta"``.
+
+.. code-block:: python
+    :caption: solvers/mysolver.py
+
+    class Solver(BaseSolver):
+        ...
+        def get_result(self):
+            return {'beta': self.beta}
+        ...
+
+
+With these methods being implemented, your solver is now ready to be run!
+
+
+Additional features
```

Comment:
This feels like too much information for a tuto, and more like a user guide of the solver features.
I would move this to another page on `The Solver object` maybe?
To me, a tuto is more step-by-step and the written file should work at the end (a bit like what jeremy did, with a screenshot of the resulting change).

Reply:
makes sense!